### ISSUE TITLE
Import support for openstack_networking_router resources

### ISSUE THREAD

**Issue by haxorof** [terraform-providers/terraform-provider-openstack#108]
### Terraform Version

0.10.4

### OpenStack Provider Version

0.2.1

### Affected Resource(s)

Please list the resources as a list, for example:

- openstack_networking_router_interface_v2
- openstack_networking_router_v2
- openstack_networking_router_route_v2

### Terraform Configuration Files

```hcl
resource "openstack_networking_router_v2" "router_1" {
  name           = "router_1"
  admin_state_up = "true"
}

resource "openstack_networking_network_v2" "network_1" {
  name           = "network_1"
  admin_state_up = "true"
}

resource "openstack_networking_subnet_v2" "subnet_1" {
  network_id = "${openstack_networking_network_v2.network_1.id}"
  cidr       = "192.168.199.0/24"
  ip_version = 4
}

resource "openstack_networking_router_interface_v2" "int_1" {
  router_id = "${openstack_networking_router_v2.router_1.id}"
  subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
}

resource "openstack_networking_router_route_v2" "router_route_1" {
  depends_on       = ["openstack_networking_router_interface_v2.int_1"]
  router_id        = "${openstack_networking_router_v2.router_1.id}"
  destination_cidr = "10.0.1.0/24"
  next_hop         = "192.168.199.254"
}
```

OpenStack provider information is only set in terraform, no environment OS prefixed variables are set.

### Expected Behavior

Router related resources imported.

### Actual Behavior

Import of resources fails because it is not supported:

```
resource openstack_networking_router_interface_v2 doesn't support import
resource openstack_networking_router_v2 doesn't support import
resource openstack_networking_router_route_v2 doesn't support import
```

### Steps to Reproduce

1. Execute `terraform apply`
2. Rename `terraform.tfstate` to `terraform.tfstate_org`
3. Import resources according to documentation but router related resources does not support import

### Possible fix

I have implemented a possible fix which I tried out but have not yet implemented any acceptance tests for these new imports:
- https://github.com/haxorof/terraform-provider-openstack/commit/1e52cba66920450850527bc64b305e8f72012384

Then it will be possible to import using the following syntax:

```
terraform import openstack_networking_router_v2.<Name> <RouterID>
terraform import openstack_networking_router_route_v2.<Name> <RouterInterfaceID>/<RouterID>/<SubnetID>/<PortID>
terraform import openstack_networking_router_interface_v2.<Name> <RouterInterfaceID>-route-<DestinationCIDR>-<NextHop>
```

**Comment by jtopjian** [comment]
@haxorof Thanks for working on this. IIRC, there was _something_ that was initially preventing a router interface from being imported, but I can't remember what. But I'm glad it's able to work now :)

The one request I have would be to keep the same import syntax of `/` to separate fields:

```
<RouterInterfaceID>/<DestinationCIDR>/<NextHop>
```

Unless there's a reason to go with `-`? As well, is there a reason to use `-route-` other than readability? In hindsight, adding those types of tags to other resources might have been helpful, but it would be a significant change to retrofit that.

With regard to the acceptance test, I can work on that part.

**Comment by haxorof** [comment]
@jtopjian Thanks for your feedback on this, I expected nothing else. :) I went for a very pragmatic approach to create the imports. 

So regarding the format for the route I intentionally took an easy path and just looked at how the ID was generated in the tfstate file. This because I did not want to spend time on it initially until I got some feedback on it as you gave me now. One reason for using `-` is because the CIDR consists of an `/` (e.g `10.1.0.0/32`). A suggestion I had which might not be that great is to replace the `-` with `#` in the CIDR and then it can be separated with slash (ex. `3e4a9105-7b10-4186-a053-dc9373bba0f4/10.0.1.0#24/192.168.199.254`). 

Please let me know if you have any suggestions on this and I fix it.

Cheers!

**Comment by jtopjian** [comment]
@haxorof ahh you are correct. The ID you're using matches the route resource ID. In that case, it works for me :)

With regard to the `openstack_networking_router_interface_v2` import, I was digging into the API, playing around with some code, and I think I was able to get it to work by only specifying a single port id. If you can, take a look at #112 and let me know what you think. 

It does not have the ability to import via subnet, but I think that's acceptable since I _think_ attaching an interface via subnet ID is just a shortcut to specifying a port ID. It's possible to get the subnet ID via the router interface port's fixed IP information, but if there's an odd case where there is more than one fixed IP, it's possible the wrong subnet will be matched. Maybe I'm wrong, though. Let me know your thoughts on this. If it's preferable to support importing via subnet, I'm open to looking into that.

I also made a quick PR for a router import (#111). I wasn't expecting it to work so easily, but it did.

If you wanted to create a PR for the route support using the pattern of #111 and #112, go for it. Also, if you wanted to try to do the import without the dedicated import function, see [this resource](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_compute_floatingip_associate_v2.go) for an example of how to integrate it into the "Read" function. I am fine with either approach.

The acceptance tests are based off of the "basic" test of the "resource" test file. For the route acceptance test, it would use the `testAccNetworkingV2RouterRoute_create` configuration.

To run the test, do:

```shell
$ cd ~/go/src/github.com/terraform-providers/terraform-provider-openstack
$ TF_LOG=DEBUG make testacc TEST=./openstack TESTARGS="TestAccNetworkingV2RouterRoute_importBasic"
```

Your `openrc` file will have to have some extra environment variables:

```
export OS_USERNAME=demo
export OS_PASSWORD=password
export OS_AUTH_URL=http://127.0.0.1:5000/v3
export OS_PROJECT_NAME=demo
export OS_USER_DOMAIN_NAME=Default
export OS_PROJECT_DOMAIN_NAME=Default
export OS_IDENTITY_API_VERSION=3

# Extra for acceptance tests
export OS_IMAGE_NAME=cirros
export OS_IMAGE_ID=818e3ca3-f505-47f4-a1b1-e4fa745ccd76
export OS_NETWORK_ID=4d3676c0-e5df-4c69-94f9-36b8e8feaefd      
export OS_EXTGW_ID=025d1f15-5d33-44e0-9a08-837254168fb6
export OS_POOL_NAME=public
export OS_FLAVOR_ID=99
export OS_FLAVOR_ID_RESIZE=98
```

For the route test, those extra environment variables can be set to fake values since they aren't used for the route tests specifically.

I'm happy to help if you run into any problems or like I mentioned before, I can just write the test. :)

**Comment by haxorof** [comment]
@jtopjian Much appreciated for your guidance on this! I had a look at the PRs and #111 #112 and it looks great to me and much more cleaner implementation than my attempt. :)

I will have a go on the route support now when you shown be how things are done right and let you know if I bump into problems. :) I am both new in golang and not really familiar to the Terraform implementation so I expect to do mistakes but want to learn and contribute.

**Comment by jtopjian** [comment]
>  I am both new in golang and not really familiar to the Terraform implementation so I expect to do mistakes but want to learn and contribute.

I hope I didn't step on your feet with #111 and #112. I wasn't expecting either to work so easily.

We're happy to help, so feel free to post questions (no question is too basic or simple) and code :)

**Comment by haxorof** [comment]
No worries! I am not emotionally attached to the code I write and only interested in reaching the our common goal meaning making this OpenStack Terraform provider great! :)

Will see if I get some time to look at the route import implementation today. Otherwise I will probably look at it later this week.

**Comment by haxorof** [comment]
@jtopjian I just noticed when I tested the latest merged commits here towards my OpenStack cloud provider that the subnet ID is not set in the tfstate file when I import the router interface using the port ID. The test sequence I did right now was:
1. Use the configuration from the first message in the issue.
2. Execute `terraform apply`
3. Manually remove the router interface part in the tfstate file.
4. Verify that `terraform plan` show that it is planning to create the router interface
5. Fetch the port ID using openstack CLI for `router_1`
6. Execute `terraform import openstack_networking_router_interface_v2.int_1 <portID>`
7. Execute `terraform plan`

On step 7 I got this result:
```
"openstack_networking_router_interface_v2.int_1": {
    "type": "openstack_networking_router_interface_v2",
    "depends_on": [],
    "primary": {
        "id": "5d9cb479-2a91-48a5-b3cd-f7e90851c4e6",
        "attributes": {
            "id": "5d9cb479-2a91-48a5-b3cd-f7e90851c4e6",
            "port_id": "5d9cb479-2a91-48a5-b3cd-f7e90851c4e6",
            "region": "Sto2",
            "router_id": "91e8be34-d665-4d28-a0da-42b322d235ec"
        },
        "meta": {},
        "tainted": false
    },
    "deposed": [],
    "provider": "openstack"
},
```
I also enable som debug logging when I executed the import: https://gist.github.com/haxorof/28376bd244a20301836c7ce45c40c444

What can cause this difference in behavior for me and you? I see in the the subnet ID in the debug log during the import.

Cheers!

**Comment by jtopjian** [comment]
That's very odd. Can you do the following:

```shell
$ OS_DEBUG=1 TF_LOG=DEBUG terraform import ...
```

And post the output?

**Comment by haxorof** [comment]
Updated the gist I previously posted now. Please have a look.

**Comment by jtopjian** [comment]
OK, everything looks correct.

> I just noticed when I tested the latest merged commits here towards my OpenStack cloud provider that the subnet ID is not set in the tfstate file when I import the router interface using the port ID.

This is expected behavior. I see that I was unclear in my explanation. It's not possible to import via subnet ID _and_ the subnet is not filled in when you import by port ID. The reason is the same: the subnet ID cannot be confidently determined if the port has more than one fixed IP.

> I see in the the subnet ID in the debug log during the import.

Yes, this is coming from the fixed IP. It's possible to pull to use the subnet that you see, but I'm hesitant.

What happens if you import and then add the resource to your terraform configuration with the subnet filled in? So the steps would be:

1. `terraform import ...`
2. `vi main.tf`
3. Create a router interface resource based on the port used in step 1 and with the subnet filled in

**Comment by haxorof** [comment]
> What happens if you import and then add the resource to your terraform configuration with the subnet filled in?

Now I tried to follow the steps but I cannot import the router interface with Terraform if it is not already defined in my Terraform configuration. So I did a test using hard-coded ID to exactly match what I created in OpenStack but still it wants to recreate it if I first run import and then plan directly after:
```hcl
resource "openstack_networking_router_interface_v2" "int_test" {
  router_id = "d0d7120f-2fc9-4d23-86e9-bab41ab1e890"
  subnet_id = "dce8537f-b82b-44cc-88fc-55dcc6271ac3"
  port_id = "84993431-2883-470b-920b-4f25d5c30c3f"
}
```
Ran plan directly after import resulted in this:
```
Terraform will perform the following actions:

-/+ openstack_networking_router_interface_v2.int_test (new resource required)
      id:        "84993431-2883-470b-920b-4f25d5c30c3f" => <computed> (forces new resource)
      port_id:   "84993431-2883-470b-920b-4f25d5c30c3f" => "84993431-2883-470b-920b-4f25d5c30c3f"
      region:    "Sto2" => <computed>
      router_id: "d0d7120f-2fc9-4d23-86e9-bab41ab1e890" => "d0d7120f-2fc9-4d23-86e9-bab41ab1e890"
      subnet_id: "" => "dce8537f-b82b-44cc-88fc-55dcc6271ac3" (forces new resource)


Plan: 1 to add, 0 to change, 1 to destroy.
```
Is it still something more I do not understand here? :)

Now over to the router routes.. :) I have tried to implement the import in the read function but I would need you to have a look at it if this is something what you are looking for. Please tell me if I have done it wrong or shall correct something:
- https://github.com/haxorof/terraform-provider-openstack/commit/3e14ed5e6deb450fd21165aab1cb43b805aa9418

Cheers!

**Comment by jtopjian** [comment]
> Is it still something more I do not understand here? 

Nope, this my mistake. `subnet_id` has `ForceNew` set. So there are three options:

1. Turn off `ForceNew` for `subnet_id` (not a good idea)
2. `subnet_id` is simply not supported for importing.
3. Modify the import functionality to take both a subnet ID and port ID.

At the moment, I'm leaning on 2 because as far as I can tell, the `subnet_id` for a router interface is only a helper to _create_ the interface and shouldn't be used as a "source of truth" so to speak since it's not possible to get the subnet ID of a router interface via an API call.

But let me look into that further. 

As well, let me know if there's a reason that `subnet_id` _must_ be set when importing an interface. Are there other resources that rely on that information or can/do they get their subnet information from another resource (such as `openstack_networking_subnet_v2`)?

I'll take a look at the routes stuff when I get a moment.

**Comment by haxorof** [comment]
What I actually looking at is to import existing OpenStack infrastructure and after I have done that Terraform shall understand that everything is up-to-date. :) Might be that I can leave out specifying the `subnet_id` but will have to do some more testing. I just saw that the new version was released with #111 and #112 so I will do some more importing on more complex setups. :) 

> At the moment, I'm leaning on 2 ...

Another possibility would maybe be to allow import functionality to take both port ID and optionally subnet ID? Then it would be possible to do the import in two different ways:
```
terraform import openstack_networking_router_interface_v2.int_test 84993431-2883-470b-920b-4f25d5c30c3f
```
... or ...
```
terraform import openstack_networking_router_interface_v2.int_test 84993431-2883-470b-920b-4f25d5c30c3f/dce8537f-b82b-44cc-88fc-55dcc6271ac3
```

If you want maybe you can skip thinking of this related to the `subnet_id` for now and maybe I can later on maybe create a new issue on that if I cannot work around it. How does that sound?

**Comment by jtopjian** [comment]
@haxorof I opened #119 which will set the `subnet_id` when it can be confidently determined. I think this is an appropriate compromise for now. If someone finds they're in a situation where this doesn't work the way they'd like, I'd be interested in seeing what they're doing and modify accordingly.

My previous concern was that a port could have multiple fixed IPs, each with their own `subnet_id` set, and therefore we wouldn't know which `subnet_id` to use. But after reviewing the neutron code which performs the [`add_router_interface`](https://github.com/openstack/neutron/blob/f0977e17b1b79c951cdac945da1a098b3bc67269/neutron/db/l3_db.py#L766) action, only 1 fixed IP is being set.

With regard to the router routes, I check out your code, tested it, and it looks good. I'd recommend the following minor updates:

1. Make a new branch of your code when you submit a PR instead of using your `master` branch. There's technically nothing wrong with `master`, it's just good practice.
2. Instead of doing `len(d.Id())`, do `d.Id() == ""`. The former can be misinterpreted that `d.Id()` is an array/slice.
3. For the `testAccCheckNetworkingV2RouterRouteDestroy` function, it might be better to check and see if the route doesn't exist. Certainly if the router doesn't exist, then the routes don't exist :smile: but checking for the actual route might be better. Something like [this](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_networking_router_route_v2.go#L74-L80) should be applicable.

Again, those are minor - overall, really good work :)

**Comment by haxorof** [comment]
@jtopjian thanks for the subnet fix you did and the feedback you given me. I will fix the stuff you mention and I think I dreamt fixing the `testAccCheckNetworkingV2RouterRouteDestroy` function because right now it is a just copy of a the function in the router test. :) I will do a PR when I am finished today or tomorrow.

**Comment by jtopjian** [comment]
I think everything has been done in this issue, so I'm going to close it. 

@haxorof thank you very much for all of your help and input with this.

**Comment by rawbertp** [comment]
Is this supposed to work with Terraform v0.11.0 / provider.openstack v0.2.1?

The reason why I am asking is because I'm still getting this error:

```
$ terraform import -var-file=cfg/uat.tfvars openstack_networking_router_interface_v2.router_int_if_net_sens 5828764a-dc02-4c91-8095-353f0d1562fc
openstack_networking_router_interface_v2.router_int_if_net_sens: Importing from ID "5828764a-dc02-4c91-8095-353f0d1562fc"...
Error importing: 1 error(s) occurred:

* openstack_networking_router_interface_v2.router_int_if_net_sens (import id: 5828764a-dc02-4c91-8095-353f0d1562fc): import openstack_networking_router_interface_v2.router_int_if_net_sens (id: 5828764a-dc02-4c91-8095-353f0d1562fc): resource openstack_networking_router_interface_v2 doesn't support import
```

**Comment by fatmcgav** [comment]
@rawbertp This fix was merged in after the 0.2.1 release, so you'll need to upgrade to 0.2.2 or newer.
